File: php/tournament/Tournament.php
<?php

define('TABLE_HEADER', "Team                           | MP |  W |  D |  L |  P");

class Tournament
{
    public function tally($scores)
    {
        $raw_results = explode("\n", $scores);
        $results = array_reduce($raw_results, 'reduce_results', array());
        usort($results, 'sort_results');
        return array_reduce($results, 'pretty_print', TABLE_HEADER);
    }
}

function pretty_print($acc, $result)
{
    return $acc . "\n" . $result->pp();
}

function sort_results($a, $b)
{
    if ($a->p == $b->p) {
        return $a->name < $b->name ? -1 : 1;
    }

    return $a->p < $b->p ? 1 : -1;
}

function reduce_results($teams, $result)
{
    trim($result);
    if ($result == "") {
        return $teams;
    }

    $parts = explode(";", $result);
    $t1_name = $parts[0];
    $t2_name = $parts[1];
    $outcome = $parts[2];

    $t1 = $teams[$t1_name] ?? new TeamResult($t1_name);
    $t2 = $teams[$t2_name] ?? new TeamResult($t2_name);

    switch ($outcome) {
        case 'win':
            $t1->add_win();
            $t2->add_loss();
            break;

        case 'loss':
            $t1->add_loss();
            $t2->add_win();
            break;

        default:
            $t1->add_draw();
            $t2->add_draw();
    }

    $teams[$t1->name] = $t1;
    $teams[$t2->name] = $t2;

    return $teams;
}

class TeamResult
{
    public function __construct($name)
    {
        $this->name = $name;
        $this->mp = 0;
        $this->w = 0;
        $this->d = 0;
        $this->l = 0;
        $this->p = 0;
    }

    public function pp()
    {
        $padded_name = str_pad($this->name, 31);
        return "{$padded_name}|  {$this->mp} |  {$this->w} |  {$this->d} |  {$this->l} |  {$this->p}";
    }

    public function add_win()
    {
        $this->mp++;
        $this->w++;
        $this->p += 3;
    }

    public function add_loss()
    {
        $this->mp++;
        $this->l++;
    }

    public function add_draw()
    {
        $this->mp++;
        $this->d++;
        $this->p++;
    }
}

File: ruby/tournament/tournament.rb
class Result
  TEMPLATE = '|  %<played>s |  %<wins>s |  %<draws>s |  %<losses>s |  %<points>s'

  private_constant :TEMPLATE

  private

  attr_writer :played, :wins, :draws, :losses, :points

  def initialize
    @played = 0
    @wins = 0
    @draws = 0
    @losses = 0
    @points = 0
  end

  def add_win
    self.wins += 1
    self.points += 3
  end

  def add_draw
    self.draws += 1
    self.points += 1
  end

  def add_loss
    self.losses += 1
  end

  public

  attr_reader :played, :wins, :draws, :losses, :points

  def save(match_issue)
    self.played += 1
    send('add_%s' % match_issue)
  end

  def to_s
    TEMPLATE % {
      played: played,
      wins: wins,
      draws: draws,
      losses: losses,
      points: points
    }
  end
end

class Tournament
  REVERSE_MATCH_ISSUE = {
    win: :loss,
    loss: :win,
    draw: :draw
  }.freeze

  TABLE_TEMPLATE = "%<title>s| MP |  W |  D |  L |  P\n%<content>s"
  TABLE_TITLE = 'Team                           '

  FIELD_SEPARATOR = ';'

  private_constant :REVERSE_MATCH_ISSUE, :TABLE_TITLE, :TABLE_TEMPLATE, :FIELD_SEPARATOR

  def self.tally(raw_results)
    TABLE_TEMPLATE % { title: TABLE_TITLE, content: new(raw_results).report }
  end

  private

  attr_accessor :results

  def initialize(raw_results)
    @results = raw_results.lines(chomp: true).each_with_object({}) do |result, results|
      next if result.empty?

      team_1, team_2, match_issue = result.split(FIELD_SEPARATOR)
      results[team_1] = Result.new unless results[team_1]
      results[team_2] = Result.new unless results[team_2]
      results[team_1].save(match_issue.to_sym)
      results[team_2].save(REVERSE_MATCH_ISSUE[match_issue.to_sym])
    end
  end

  def sort
    results.sort_by { |team, result| [-result.points, team] }
  end

  public

  def report
    sort.sum('') do |team, result|
      "%<team>s%<padding>s%<result>s\n" % {
        team: team,
        padding: ' ' * (TABLE_TITLE.length - team.length),
        result: result
      }
    end
  end
end
